src/pal/src/cruntime/misctls.cpp

```diff
@@ -150,250 +164,155 @@ See MSDN for more information.
-    /* Get the per-thread buffer from the thread structure. */
-    pThread = InternalGetCurrentThread();
-
-    lpStartOfReturnBuffer = lpReturnBuffer = pThread->crtInfo.ECVTBuffer;
-
-    /* Sanity checks */
-    if ( !dec || !sign )
+    if (count > ECVT_MAX_COUNT_SIZE)
     {
-        ERROR( "dec and sign have to be valid pointers.\n" );
-        *lpReturnBuffer = '\0';
-        goto done;
-    }
-    else
-    {
-        *dec = *sign = 0;
+        count = ECVT_MAX_COUNT_SIZE;
     }
 
-    if ( value < 0.0 )
-    {
-        *sign = 1;
-    }
+    const int SCALE_NAN = (int)0x80000000;
+    const int SCALE_INF = 0x7FFFFFFF;
 
-    if ( count > ECVT_MAX_COUNT_SIZE )
+    if (GetExponent(value) == 0x7ff)
     {
-        count = ECVT_MAX_COUNT_SIZE;
+        *dec = GetMantissa(value) != 0 ? SCALE_NAN : SCALE_INF;
+        *sign = GetSign(value);
+        lpStartOfReturnBuffer[0] = '\0';
+        return lpStartOfReturnBuffer;
     }
 
-    /* Get the string to work with. */
-    sprintf_s( TempBuffer, sizeof(TempBuffer), FORMAT_STRING, value );
+#ifndef __APPLE__
+    // OSX doesn't support evct_r. it support evct but we cannot use it either as it is not thread safe
+    ecvt_r(value, count, dec, sign, lpStartOfReturnBuffer, ECVT_MAX_BUFFER_SIZE);
+    return lpStartOfReturnBuffer;
+#else // __APPLE__
+
+    CHAR TempBuffer[ECVT_MAX_BUFFER_SIZE];
+   
+    *dec = *sign = 0;
 
-    /* Check to see if value was a valid number. */
-    if ( strcmp( "NaN", TempBuffer ) == 0 || strcmp( "-NaN", TempBuffer ) == 0 )
+    if (value < 0.0)
     {
-        TRACE( "value was not a number!\n" );
-        if (strcpy_s( lpStartOfReturnBuffer, ECVT_MAX_BUFFER_SIZE, "1#QNAN0" ) != SAFECRT_SUCCESS)
-        {
-            ERROR( "strcpy_s failed!\n" );
-            *lpStartOfReturnBuffer = '\0';
-            goto done;
-        }
-
-        *dec = 1;
-        goto done;
+        *sign = 1;
     }
-
-    /* Check to see if it is infinite. */
-    if ( strcmp( "Inf", TempBuffer ) == 0 || strcmp( "-Inf", TempBuffer ) == 0 )
+    
+    if (value == 0.0)
     {
-        TRACE( "value is infinite!\n" );
-        if (strcpy_s( lpStartOfReturnBuffer, ECVT_MAX_BUFFER_SIZE, "1#INF00" ) != SAFECRT_SUCCESS)
+        for (int j = 0; j < count; j++)
         {
-            ERROR( "strcpy_s failed!\n" );
-            *lpStartOfReturnBuffer = '\0';
-            goto done;
+            lpStartOfReturnBuffer[j] = '0';
         }
+        lpStartOfReturnBuffer[count] = '\0';
+        return lpStartOfReturnBuffer;
+    } 
+    
+    int tempBufferLength = snprintf(TempBuffer, ECVT_MAX_BUFFER_SIZE, "%.40e", value);
+    _ASSERTE(tempBufferLength > 0 && ECVT_MAX_BUFFER_SIZE > tempBufferLength);
+    
+    //
+    // Calculate the exponent value
+    //
 
-        *dec = 1;
-        if ( *TempBuffer == '-' )
-        {
-            *sign = 1;
-        }
-        goto done;
+    int exponentIndex = tempBufferLength - 1;
+    while (TempBuffer[exponentIndex] != 'e' && exponentIndex > 0)
```

Comment:
nit: It might not matter here, but isn't `strrchr` more efficient, since it might use vectorization in some cases?